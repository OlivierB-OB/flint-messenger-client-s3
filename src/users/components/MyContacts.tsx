import Box from '@material-ui/core/Box';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Drawer from '@material-ui/core/Drawer';
import React from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { makeCreateConversation } from '../../conversations/actions/makeCreateConversation';
import { hideUsersList } from '../actions/hideUsersList';
import { IUserInfo, IUsersStateStatus } from '../types';
import { ContactListItem } from './ContactListItem';
import { Loading } from '../../layout/components/Loading';

export interface IContactListProps {
  status: IUsersStateStatus;
  show: boolean;
  list: IUserInfo[];
  hideContactList: () => void;
  createConversation: (targetId: string) => void;
}

export function ContactList(props: IContactListProps) {
  const { status, list, show, hideContactList, createConversation } = props;
  const loading = status === 'unavailable' ? <Loading /> : null;
  return (
    <Drawer anchor="left" open={show} onClose={hideContactList}>
      <Box style={{ minWidth: '300px' }}>
        {loading}
        <List>
          {loading || list.length ? null : (
            <ListItem>
              <ListItemText primary="no contact available..." />
            </ListItem>
          )}
          {list.map((user) => (
            <ListItem button onClick={() => createConversation(user.uid)} key={user.uid}>
              <ContactListItem info={user} />
            </ListItem>
          ))}
        </List>
      </Box>
    </Drawer>
  );
}

const mapStateToProps = ({ users }: IAppState) => ({
  status: users.status,
  show: users.show,
  list: users.list,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  hideContactList: () => dispatch(hideUsersList()),
  createConversation: (targetId: string) => dispatch(makeCreateConversation(targetId)),
});

export const MyContacts = connect(mapStateToProps, mapDispatchToProps)(ContactList);
