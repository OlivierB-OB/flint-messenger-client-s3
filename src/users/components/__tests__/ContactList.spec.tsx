import ListItem from '@material-ui/core/ListItem';
import React from 'react';
import { shallow } from 'enzyme';
import { ContactList } from '../MyContacts';

describe('ContactList', () => {
  it('should appear as loading if users service is unavailable', async () => {
    const list: any = [];
    function createConversation() {}
    const component = shallow(
      <ContactList status="unavailable" list={list} createConversation={createConversation} />,
    ).dive();
    expect(component).toMatchSnapshot();
  });

  it('should display a message if no contact', async () => {
    const list: any = [];
    function createConversation() {}
    const component = shallow(
      <ContactList status="ready" list={list} createConversation={createConversation} />,
    ).dive();
    expect(component).toMatchSnapshot();
  });

  it('should display the contact', async () => {
    const list: any = [
      {
        _id: '123',
      },
      {
        _id: '124',
      },
    ];
    function createConversation() {}
    const component = shallow(
      <ContactList status="ready" list={list} createConversation={createConversation} />,
    ).dive();
    expect(component).toMatchSnapshot();
  });

  it('should allow selecting a contact to create a new conversation', async () => {
    const list: any = [
      {
        _id: '123',
      },
      {
        _id: '124',
      },
    ];
    const createConversation = jest.fn();
    const component = shallow(
      <ContactList status="ready" list={list} createConversation={createConversation} />,
    ).dive();
    component.find(ListItem).at(0).simulate('click');
    expect(createConversation).toHaveBeenCalledWith('123');
  });
});
