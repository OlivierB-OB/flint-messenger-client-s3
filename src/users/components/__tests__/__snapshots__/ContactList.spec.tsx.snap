// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactList should appear as loading if users service is unavailable 1`] = `
<div
  className="MuiBox-root MuiBox-root-1"
  style={
    Object {
      "minWidth": "300px",
    }
  }
>
  <Loading />
  <WithStyles(ForwardRef(List)) />
</div>
`;

exports[`ContactList should display a message if no contact 1`] = `
<div
  className="MuiBox-root MuiBox-root-2"
  style={
    Object {
      "minWidth": "300px",
    }
  }
>
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))>
      <WithStyles(ForwardRef(ListItemText))
        primary="no contact available..."
      />
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</div>
`;

exports[`ContactList should display the contact 1`] = `
<div
  className="MuiBox-root MuiBox-root-3"
  style={
    Object {
      "minWidth": "300px",
    }
  }
>
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      key="123"
      onClick={[Function]}
    >
      <ContactListItem
        info={
          Object {
            "_id": "123",
          }
        }
      />
    </WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      key="124"
      onClick={[Function]}
    >
      <ContactListItem
        info={
          Object {
            "_id": "124",
          }
        }
      />
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</div>
`;
